### PR TITLE
AV-219152 Moving HTTPRoute Validation to Graph layer

### PR DESCRIPTION
**Pull request by pkoshtavmware**
AV-219152 Moving HTTPRoute Validation to Graph layer

### CODE REVIEW

**Comment by pkoshtavmware** [comment]
FT run result:
Screenshot 2024-09-30 at 3 47 09 PM

**Comment by pkoshtavmware** [comment]
UT run after changes:
`ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/k8stest	30.422s	coverage: 8.8% of statements in ./...
coverage: 26.3% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/integrationtest	499.099s	coverage: 26.3% of statements in ./...
coverage: 34.7% of statements in ./...
coverage: 8.8% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/k8stest	30.402s	coverage: 8.8% of statements in ./...
coverage: 26.3% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/integrationtest	499.061s	coverage: 26.3% of statements in ./...
coverage: 39.0% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/ingresstests	1101.039s	coverage: 39.0% of statements in ./...
coverage: 38.3% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/evhtests	2039.361s	coverage: 38.3% of statements in ./...
coverage: 37.6% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/evhtests	1846.792s	coverage: 37.6% of statements in ./...
coverage: 28.3% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/dedicatedevhtests	692.522s	coverage: 28.3% of statements in ./...
coverage: 28.6% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/dedicatedevhtests	788.508s	coverage: 28.6% of statements in ./...
coverage: 33.0% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/oshiftroutetests	1313.063s	coverage: 33.0% of statements in ./...
coverage: 4.0% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/bootuptests	20.154s	coverage: 4.0% of statements in ./...
coverage: 8.2% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/multicloudtests	170.445s	coverage: 8.2% of statements in ./...
coverage: 19.2% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/advl4tests	46.767s	coverage: 19.2% of statements in ./...
coverage: 21.2% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/namespacesynctests	152.395s	coverage: 21.2% of statements in ./...
coverage: 22.3% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/servicesapitests	79.008s	coverage: 22.3% of statements in ./...
coverage: 23.7% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/npltests	116.831s	coverage: 23.7% of statements in ./...
coverage: 30.4% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/dedicatedvstests	187.484s	coverage: 30.4% of statements in ./...
coverage: 25.0% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/hatests	33.872s	coverage: 25.0% of statements in ./...
coverage: 13.1% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/cnitests	103.556s	coverage: 13.1% of statements in ./...
coverage: 12.8% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/cnitests	27.278s	coverage: 12.8% of statements in ./...
coverage: 21.6% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/urltests	15.238s	coverage: 21.6% of statements in ./...
	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests		coverage: 0.0% of statements
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/graphlayer	290.875s	coverage: 18.0% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/ingestion	220.582s	coverage: 2.5% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/npltests	60.323s	coverage: 15.7% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/status	300.652s	coverage: 12.6% of statements in ./...`

**Comment by pkoshtavmware** [comment]
FT run result:
Screenshot 2024-10-03 at 12 38 04 PM

**Comment by pkoshtavmware** [comment]
UT run result:
```
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/k8stest	30.418s	coverage: 8.8% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/integrationtest	499.124s	coverage: 26.3% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/ingresstests	1101.060s	coverage: 38.8% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/evhtests	2039.405s	coverage: 38.4% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/evhtests	1846.766s	coverage: 37.8% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/dedicatedevhtests	692.489s	coverage: 28.3% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/dedicatedevhtests	788.512s	coverage: 28.6% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/oshiftroutetests	1313.071s	coverage: 33.0% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/bootuptests	20.187s	coverage: 4.0% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/multicloudtests	170.451s	coverage: 8.2% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/advl4tests	46.756s	coverage: 19.1% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/namespacesynctests	152.410s	coverage: 21.2% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/servicesapitests	79.040s	coverage: 22.3% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/npltests	116.844s	coverage: 23.7% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/dedicatedvstests	187.498s	coverage: 30.2% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/hatests	33.880s	coverage: 25.0% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/cnitests	103.560s	coverage: 13.1% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/cnitests	27.259s	coverage: 12.8% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/urltests	15.204s	coverage: 21.6% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/graphlayer	290.915s	coverage: 18.0% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/ingestion	220.636s	coverage: 2.5% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/npltests	60.357s	coverage: 15.7% of statements in ./...
ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/status	300.674s	coverage: 12.7% of statements in ./...
```

File: ako-gateway-api/nodes/dequeue_ingestion.go [hunk 1]
```diff
@@ -36,6 +36,15 @@ func DequeueIngestion(key string, fullsync bool) {
 	if !valid {
 		return
 	}
+	if objType == lib.HTTPRoute {
+		httpRoute, err := akogatewayapilib.AKOControlConfig().GatewayApiInformers().HTTPRouteInformer.Lister().HTTPRoutes(namespace).Get(name)
+		if err == nil {
+			utils.AviLog.Debugf("key: %s, msg: Successfully retrieved the HTTPRoute object %s", key, name)
+			if !IsHTTPRouteValid(key, httpRoute) {
```
Comment (akshayhavile) [inline ako-gateway-api/nodes/dequeue_ingestion.go:43]: @pkoshtavmware: Should we include this as part of `HTTPRouteChanges` function? How transition from valid httproute to invalid HTTP route will be handled with return into picture? Are we deleting previously created objects with invalid transition?

File: tests/gatewayapitests/status/gateway_test.go [hunk 2]
```diff
@@ -1125,15 +1214,27 @@ func TestMultipleGatewaySameHostname(t *testing.T) {
...
 		},
-		Listeners: tests.GetListenerStatusV1(ports, []int32{0, 0}),
+		Listeners: tests.GetListenerStatusV1(ports, []int32{0, 0}, false, true),
 	}
-	expectedStatus.Listeners[0].Conditions[0].Reason = string(gatewayv1.GatewayReasonListenersNotValid)
+	expectedStatus.Listeners[0].Conditions[0].Reason = string(gatewayv1.ListenerReasonInvalid)
 	expectedStatus.Listeners[0].Conditions[0].Status = metav1.ConditionFalse
+	expectedStatus.Listeners[0].Conditions[0].Message = "Hostname overlaps or is same as an existing gateway hostname"
+
+	expectedStatus.Listeners[0].Conditions[1].Reason = string(gatewayv1.ListenerReasonInvalid)
```
Comment (akshayhavile) [inline tests/gatewayapitests/status/gateway_test.go:1235]: Do we have to check other condition (listener being invalid) when original issue is overlapping hostnames?

File: ako-gateway-api/status/status.go [hunk 3]
```diff
@@ -80,7 +84,7 @@ func BulkUpdate(key string, objectType string, options []status.StatusOptions) e
 	return nil
 }
 
-func Record(key string, obj runtime.Object, status *Status) {
+func Record(key string, obj runtime.Object, status *Status) (runtime.Object, error) {
```
Comment (akshayhavile) [inline ako-gateway-api/status/status.go:87]: @pkoshtavmware: I think we are not using return value `runtime.Object` anywhere. Can we take it out (in nested calls also)?

File: ako-gateway-api/k8s/validator.go [hunk 4]
```diff
@@ -92,47 +87,81 @@ func IsValidGateway(key string, gateway *gatewayv1.Gateway) bool {
...
 			Reason(string(gatewayv1.GatewayReasonListenersNotValid)).
-			Message(fmt.Sprintf("Gateway contains %d invalid listener(s)", invalidListenerCount)).
+			Message("Gateway does not contain any valid listener").
+			SetIn(&gatewayStatus.Conditions)
+		programmedCondition.
 			SetIn(&gatewayStatus.Conditions)
 		akogatewayapistatus.Record(key, gateway, &akogatewayapistatus.Status{GatewayStatus: gatewayStatus})
+		gateway.Status = *gatewayStatus.DeepCopy()
 		return false
+	} else if validListenerCount < len(spec.Listeners) {
```
[the remark this answers is not in the document]
Comment (pkoshtavmware) [reply ako-gateway-api/k8s/validator.go:137]: Those changes were needed here in order to test this one end to end, hence cherry-picked them here. Once they are merged to master i will rebase and these changes here will be gone

File: ako-gateway-api/k8s/validator.go [hunk 5]
```diff
@@ -245,234 +305,28 @@ func isValidListener(key string, gateway *gatewayv1.Gateway, gatewayStatus *gate
...
+		Reason(string(gatewayv1.ListenerReasonResolvedRefs)).
+		Message("All the references are valid").
+		SetIn(&gatewayStatus.Listeners[index].Conditions)
+
 	utils.AviLog.Infof("key: %s, msg: Listener %s/%s is valid", key, gateway.Name, listener.Name)
 	return true
 }
 
-func IsHTTPRouteValid(key string, obj *gatewayv1.HTTPRoute) bool {
+func IsHTTPRouteConfigValid(key string, obj *gatewayv1.HTTPRoute) bool {
```
[the remark this answers is not in the document]
Comment (pkoshtavmware) [reply ako-gateway-api/k8s/validator.go:324]: I have a ticket open for that already https://vmw-jira.broadcom.net/browse/AV-206859. Will do those changes on top of all the reordering and partial listener changes.

File: ako-gateway-api/status/status.go (hunk 3 above)
Comment (pkoshtavmware) [reply ako-gateway-api/status/status.go]: Done

File: tests/gatewayapitests/status/gateway_test.go (hunk 2 above)
Comment (pkoshtavmware) [reply tests/gatewayapitests/status/gateway_test.go:1235]: Yes. I have crosschecked with nginx implementation here.
https://github.com/nginxinc/nginx-gateway-fabric/blob/adb76bf149fb3aa98f0a85403b59c2c9d0331f21/internal/mode/static/state/conditions/conditions.go#L377 here

**Comment by pkoshtavmware** [comment]
Ut runs:
    ```
    github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests               coverage: 0.0% of statements
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/graphlayer    291.243s        coverage: 18.1% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/ingestion     220.689s        coverage: 2.7% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/npltests      60.483s coverage: 16.0% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests/status        300.773s        coverage: 12.8% of statements in ./...
```

**Review by akshayhavile — APPROVED** [review]
Few minor comments.

**Review by arihantg — APPROVED** [review]
LGTM

File: ako-gateway-api/k8s/validator.go [hunk 6]
```diff
@@ -189,32 +208,51 @@ func isValidListener(key string, gateway *gatewayv1.Gateway, gatewayStatus *gate
...
+				resolvedRefCondition.Reason(string(gatewayv1.ListenerReasonInvalidCertificateRef)).
 					Message("TLS mode or reference not valid").
 					SetIn(&gatewayStatus.Listeners[index].Conditions)
+				programmedCondition.SetIn(&gatewayStatus.Listeners[index].Conditions)
+				return false
+			}
+			name := string(certRef.Name)
+			cs := utils.GetInformers().ClientSet
+			secretObj, err := cs.CoreV1().Secrets(gateway.ObjectMeta.Namespace).Get(context.TODO(), name, metav1.GetOptions{})
+			if err != nil || secretObj == nil {
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply ako-gateway-api/k8s/validator.go:246]: will be handled in next pr.

**Pull request merged**